File: min_stack.py


class MinStack:
    _stack = None
    _min_idxs = None

    def __init__(self):
        self._stack = []
        self._min_idxs = []

    # @param x, an integer
    # @return an integer
    def push(self, x):
        self._stack.append(x)
        if not self._min_idxs or self._stack[self._min_idxs[-1]] > x:
            self._min_idxs.append(len(self._stack)-1)


    # @return nothing
    def pop(self):
        if self._min_idxs and self._min_idxs[-1] == len(self._stack) - 1:
            self._min_idxs.pop()
        self._stack.pop()


    # @return an integer
    def top(self):
        if not self._stack:
            return None
        return self._stack[-1]

    # @return an integer
    def getMin(self):
        if not self._stack:
            return None
        return self._stack[self._min_idxs[-1]]



def main():
    stack = MinStack()

    stack.push(-1)
    print stack.getMin()
    print stack.top()
    stack.push(50)
    stack.push(4)
    stack.push(-4)
    print stack.getMin()
    print stack.top()
    stack.pop()
    stack.pop()
    print stack.getMin()
    stack.pop()
    stack.pop()
    print stack.getMin()

if __name__ == '__main__':
    main()

File: src/min_stack/min_stack.cpp
//min stack
#include <iostream>
#include <cstdlib>
#include <stack>
using namespace std;
/*class MinStack {
private:
    vector<int> stack;
    vector<int> minStack;
    int index;
    int mindex;
public:
    MinStack()
    {
        vector<int> stack(100,0);
        vector<int> minStack(100,0);
        for(int i=0;i<10;i++)
        {
                stack[i]=0;
                minStack[i]=0;
        }
                index =0;
                mindex=0;
    }
    void push(int x) {
        if(mindex==0)
        {
           minStack[mindex]=x;
           mindex++;         
                    }
        else if(x<=minStack[mindex-1]){
            minStack[mindex]=x;
            mindex++;
        }
        stack[index]=x;
        index++;
    }

    void pop() {
        if(stack[index-1]==minStack[mindex-1]){
            minStack[mindex]=0;
            mindex--;
        }
        stack[index]=0;
        index--;
    }

    int top() {
        return stack[index-1];
    }

    int getMin() {
        return minStack[mindex-1];
    }
};*/
class MinStack {
    stack<int> s;
    stack<int> ms;
public:
    void push(int x) {
        s.push(x);
        if(ms.empty() || x<=ms.top()) 
            ms.push(x);
    }

    void pop() {
        if(s.empty()) return;
        if(s.top()==ms.top()) ms.pop();
        s.pop();
    }

    int top() {
        if(!s.empty())
        return s.top();
    }

    int getMin() {
        if(!ms.empty())
        return ms.top();
    }
};
int main(){
    
    MinStack ms1; 
    ms1.push(5);
    cout<<ms1.top()<<endl;
    ms1.push(10);
    cout<<ms1.top()<<endl;
    ms1.pop();
    cout<<ms1.top()<<endl;
    ms1.push(-1);
    cout<<ms1.top()<<endl;
    ms1.push(-8);
    ms1.pop();
    cout<<ms1.top()<<endl;
    cout<<ms1.getMin();
    system("pause");
    return 0;
    
    }
